JP-1154: Flag CR neighbors  #4297

mwregan2 opened the pull request:
These are the changes needed to add the capability to flag the neighbors of detected jumps.

The implementation sets flagging neighbors to be on by default but allows it to be turned off with a config parameter.

Resolves #4297 / JP-1154

mwregan2 commented:
I'm sorry about the changes to .gitignore. At first I though that I shouldn't change it but then saw what looked like visual studio files. Anyway, I removed the Pycharm directory, fixed the MIRIRampModel problem, skipped the long running test that is just for performance testing.

philhodge commented on jwst/jump/tests/test_detect_jumps.py:
```diff
@@ -82,27 +414,36 @@ def test_onecr_50_groups(setup_inputs):
-        data = np.zeros(shape=(nints, ngroups, nrows, ncols), dtype=np.float64)
+
+
         pixdq = np.zeros(shape=(nrows, ncols), dtype=np.float64)
```
The data type for `pixdq` should be `np.uint32`.  This change should also be made in test_jump_step.py.

mwregan2 replied:
changed it in all three test routines.

philhodge commented on jwst/jump/tests/test_jump_step.py:
```diff
@@ -227,3 +227,22 @@ def test_two_CRs(generate_miri_reffiles, max_cores, setup_inputs):
+        gain=ingain, readnoise=inreadnoise, deltatime=grouptime)
+    out_model = JumpStep.call(model1, override_gain=override_gain,
+                              override_readnoise=override_readnoise, maximum_cores=max_cores)
+    assert(out_model.meta.cal_step.jump == 'SKIPPED')
```
It looks as if there is no newline at the end of this file.

mwregan2 replied:
added the newline

hbushouse commented on jwst/jump/jump_step.py:
```diff
@@ -45,13 +48,15 @@ def process(self, input):
+            max_jump_to_flag_neighbors = self.max_jump_to_flag_neighbors
+            min_jump_to_flag_neighbors = self.min_jump_to_flag_neighbors
+            flag_4_neighbors = self.flag_4_neighbors
             do_yint = self.do_yintercept
```
Can remove these 2 lines, since y_int is being stripped out.

mwregan2 replied:
removed them.

hbushouse commented on jwst/jump/jump_step.py:
```diff
@@ -67,7 +72,9 @@ def process(self, input):
             # Call the jump detection routine
             result = detect_jumps(input_model, gain_model, readnoise_model,
-                                   rej_thresh, do_yint, sig_thresh, max_cores)
+                                rej_thresh, do_yint, sig_thresh, max_cores,
```
Can remove `do_yint` and `sig_thresh` from argument list, since that method is being completely ripped out.

mwregan2 replied:
Removed them here and in all calls to detect_jumps

hbushouse commented on jwst/jump/twopoint_difference.py:
```diff
@@ -30,8 +31,11 @@ def find_crs(data, group_dq, read_noise, rej_threshold, nframes):
 
-    # Create array for output median slope images
+    # Create arrays for output
     median_slopes = np.zeros((nints, nrows, ncols), dtype=np.float32)
```
As mentioned earlier, I don't think we really need median_slopes anymore.

mwregan2 replied:
This array is used within twopoint_difference. It is not longer returned or mentioned in other routines or tests.

mwregan2 replied:
needed internally but no longer is returned

mwregan2 commented on jwst/jump/jump.py:
```diff
@@ -88,10 +87,18 @@ def detect_jumps (input_model, gain_model, readnoise_model,
+    row_below_gdq = np.zeros((num_ints, num_groups, ncols), dtype=np.uint8)
+
+    yincrement = int(nrows / numslices)
+    print("yincrement ", yincrement)
```
removed

mwregan2 commented on jwst/jump/jump_step.py:
```diff
@@ -17,6 +17,9 @@ class JumpStep(Step):
         rejection_threshold = float(default=4.0,min=0) # CR rejection threshold
         maximum_cores = option('quarter', 'half', 'all', default=None) # max number of processes to create
+        flag_4_neighbors = boolean(default=True) #flag the four perpendicular neighbors of each CR
+        max_jump_to_flag_neighbors = float(default=10) #maximum jump that will trigger neighbor flagging
```
good point, I changed it.

mwregan2 commented on jwst/jump/tests/test_detect_jumps.py:
```diff
@@ -31,17 +60,312 @@ def test_onecr_10_groups(setup_inputs):
+    model1.data[1, 3, 15, 5] = 30.0
+    model1.data[1, 4, 15, 5] = 35.0
+    model1.data[1, 5, 15, 5] = 40.0
+    model1.data[1, 6, 15, 5] = 50.0
```
I changed it.

mwregan2 commented on jwst/jump/tests/test_detect_jumps.py:
```diff
@@ -31,17 +60,312 @@ def test_onecr_10_groups(setup_inputs):
+    numslices = num_cores // 2
+    yincrement = int(nrows / numslices)
+    # two segments perfect fit, second segment has twice the slope
+    #add a CR on the last row of the first slice
```
There was an error in this test that if the number of slices is 1, it would write outside the array. I added a test to skip the test if the number of slices is 1. Otherwise it works with any number of slices greater than 1.

mwregan2 commented on jwst/jump/tests/test_detect_jumps.py:
```diff
@@ -31,17 +60,312 @@ def test_onecr_10_groups(setup_inputs):
+
+def test_onecr_10_groups_fullarray(setup_inputs):
+    """"
+      A test to that has a cosmic ray in the 5th group for all pixels except row 10. In row
```
changed it.

mwregan2 commented on jwst/jump/tests/test_detect_jumps.py:
```diff
@@ -31,17 +60,312 @@ def test_onecr_10_groups(setup_inputs):
+    model1.data[1, 3, 15, 5] = 30.0
+    model1.data[1, 4, 15, 5] = 35.0
+    model1.data[1, 5, 15, 5] = 40.0
+    model1.data[1, 6, 15, 5] = 50.0
```
changed

mwregan2 commented on jwst/jump/tests/test_detect_jumps.py:
```diff
@@ -31,17 +60,312 @@ def test_onecr_10_groups(setup_inputs):
+
+def test_flagging_of_CRs_across_slice_boundaries(setup_inputs):
+    """"
+            A multiprocesssing test that has two CRs on the boundary between two
```
changed all of them. copy and paste is not good when you start with it misspelled.

mwregan2 commented on jwst/jump/tests/test_detect_jumps.py:
```diff
@@ -82,27 +414,36 @@ def test_onecr_50_groups(setup_inputs):
-    assert (4 == np.max(out_model.groupdq[0, 5, 5, 5]))
-    outdqcr = out_model.groupdq[0, 5, 5, 5]
-    np.testing.assert_allclose(4, outdqcr)
+    print(np.arange(190, 290, 5))
```
deleted

mwregan2 commented on jwst/jump/jump.py:
```diff
@@ -116,33 +128,31 @@ def detect_jumps (input_model, gain_model, readnoise_model,
-
-    # Apply the y-intercept method as a second pass, if requested
-    if do_yint:
+    log.info('Elapsed time of twopoint difference = %g sec' % elapsed)
```
you are right. They are the same. I removed the twopoint specific one.

mwregan2 commented on jwst/jump/jump.py:
```diff
@@ -116,33 +128,31 @@ def detect_jumps (input_model, gain_model, readnoise_model,
-        log.info('Executing yintercept method')
-        start = time.time()
-        yint.find_crs(data, err, gdq, times, readnoise_2d,
-                        rejection_threshold, signal_threshold, median_slopes)
```
I removed it everywhere but inside of twopoint_difference

mwregan2 commented on jwst/jump/jump_step.py:
```diff
@@ -17,6 +17,9 @@ class JumpStep(Step):
     spec = """
         rejection_threshold = float(default=4.0,min=0) # CR rejection threshold
         maximum_cores = option('quarter', 'half', 'all', default=None) # max number of processes to create
+        flag_4_neighbors = boolean(default=True) #flag the four perpendicular neighbors of each CR
```
fixed

mwregan2 commented:
Is there something preventing the pull request from getting merged?

hbushouse commented:
> Is there something preventing the pull request from getting merged?

Yes, there are merge conflicts for `test_detect_jumps.py`. Can you fix those?

Also wasn't sure if you were going to include documentation updates in this PR or a separate one. If you're going to do it in a separate PR, then we can merge this one (once the conflict is fixed).

mwregan2 commented:
Howard,
  I saw the flagged conflict but I don’t think it is correct. I’ll look at it again.
  I was planning on doing the doc updates in a separate PR. Since, I assume they will need a review.
Mike


From: Howard Bushouse <notifications@github.com>
Reply-To: spacetelescope/jwst <reply@reply.github.com>
Date: Wednesday, February 19, 2020 at 1:34 PM
To: spacetelescope/jwst <jwst@noreply.github.com>
Cc: Michael Regan <mregan@stsci.edu>, Author <author@noreply.github.com>
Subject: Re: [spacetelescope/jwst] JP-1154: Flag CR neighbors #4297 (#4434)


Is there something preventing the pull request from getting merged?

Yes, there are merge conflicts for test_detect_jumps.py. Can you fix those?

Also wasn't sure if you were going to include documentation updates in this PR or a separate one. If you're going to do it in a separate PR, then we can merge this one (once the conflict is fixed).

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/spacetelescope/jwst/pull/4434?email_source=notifications&email_token=AF6KJLTZPFRNH3YHCHTL5ELRDV3SBA5CNFSM4KD6HEI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMI6LSI#issuecomment-588375497>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AF6KJLRPWPDDGDWRG523KVTRDV3SBANCNFSM4KD6HEIQ>.

dmggh commented:
I'm glad to learn that doc updates will be in a separate PR, as I suspect that will help me better understand these code updates.

________________________________
From: mwregan2 <notifications@github.com>
Sent: Wednesday, February 19, 2020 1:36 PM
To: spacetelescope/jwst
Cc: David Grumm; Review requested
Subject: Re: [spacetelescope/jwst] JP-1154: Flag CR neighbors #4297 (#4434)

Howard,
I saw the flagged conflict but I don’t think it is correct. I’ll look at it again.
I was planning on doing the doc updates in a separate PR. Since, I assume they will need a review.
Mike


From: Howard Bushouse <notifications@github.com>
Reply-To: spacetelescope/jwst <reply@reply.github.com>
Date: Wednesday, February 19, 2020 at 1:34 PM
To: spacetelescope/jwst <jwst@noreply.github.com>
Cc: Michael Regan <mregan@stsci.edu>, Author <author@noreply.github.com>
Subject: Re: [spacetelescope/jwst] JP-1154: Flag CR neighbors #4297 (#4434)


Is there something preventing the pull request from getting merged?

Yes, there are merge conflicts for test_detect_jumps.py. Can you fix those?

Also wasn't sure if you were going to include documentation updates in this PR or a separate one. If you're going to do it in a separate PR, then we can merge this one (once the conflict is fixed).

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/spacetelescope/jwst/pull/4434?email_source=notifications&email_token=AF6KJLTZPFRNH3YHCHTL5ELRDV3SBA5CNFSM4KD6HEI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMI6LSI#issuecomment-588375497>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AF6KJLRPWPDDGDWRG523KVTRDV3SBANCNFSM4KD6HEIQ>.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub<https://github.com/spacetelescope/jwst/pull/4434?email_source=notifications&email_token=AA4MFDX76MAWVMHWL4BIISTRDV3ZPA5CNFSM4KD6HEI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMI6TZA#issuecomment-588376548>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AA4MFDWWQNQQQDNRCAQSYMLRDV3ZPANCNFSM4KD6HEIQ>.

mwregan2 commented:
Howard,
  Git got confused with the fact that I added some if tests just in front of the change in Master. I had already updated to RampModel. I resolved the conflict and merged it. Since I don’t have merge authority, that failed but the CI steps are running now and no conflict is listed.
Mike


From: Howard Bushouse <notifications@github.com>
Reply-To: spacetelescope/jwst <reply@reply.github.com>
Date: Wednesday, February 19, 2020 at 1:34 PM
To: spacetelescope/jwst <jwst@noreply.github.com>
Cc: Michael Regan <mregan@stsci.edu>, Author <author@noreply.github.com>
Subject: Re: [spacetelescope/jwst] JP-1154: Flag CR neighbors #4297 (#4434)


Is there something preventing the pull request from getting merged?

Yes, there are merge conflicts for test_detect_jumps.py. Can you fix those?

Also wasn't sure if you were going to include documentation updates in this PR or a separate one. If you're going to do it in a separate PR, then we can merge this one (once the conflict is fixed).

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/spacetelescope/jwst/pull/4434?email_source=notifications&email_token=AF6KJLTZPFRNH3YHCHTL5ELRDV3SBA5CNFSM4KD6HEI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMI6LSI#issuecomment-588375497>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AF6KJLRPWPDDGDWRG523KVTRDV3SBANCNFSM4KD6HEIQ>.

hbushouse commented:
OK, once the CI tests are complete I'll merge.

mwregan2 commented:
I do I find out what the problem is that caused the two Jenkins tests to fail?

hbushouse commented:
Looks like something in `test_detect_jumps`is causing a hard error in the Travis CI tests.

hbushouse commented:
You can try running the unit tests locally to see if the same error occurs (i.e just do a "pytest" in the test directory of your jump source code)

mwregan2 commented:
Howard,
  I saw that but I just reran the tests locally and they all passed
Mike.

From: Howard Bushouse <notifications@github.com>
Reply-To: spacetelescope/jwst <reply@reply.github.com>
Date: Wednesday, February 19, 2020 at 2:39 PM
To: spacetelescope/jwst <jwst@noreply.github.com>
Cc: Michael Regan <mregan@stsci.edu>, Author <author@noreply.github.com>
Subject: Re: [spacetelescope/jwst] JP-1154: Flag CR neighbors #4297 (#4434)


Looks like something in test_detect_jumpsis causing a hard error in the Travis CI tests.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/spacetelescope/jwst/pull/4434?email_source=notifications&email_token=AF6KJLXEF52GMHRV5DJZUNLRDWDHFA5CNFSM4KD6HEI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMJGZ3Q#issuecomment-588410094>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AF6KJLVU234QUWBX6N4A7KLRDWDHFANCNFSM4KD6HEIQ>.

hbushouse commented:
Huh, that's not good (or normal).

mwregan2 commented:
Howard,
  I may have found something. It looks like I’ve got MIRIRampModel in my GitHub source.
Mike


From: Howard Bushouse <notifications@github.com>
Reply-To: spacetelescope/jwst <reply@reply.github.com>
Date: Wednesday, February 19, 2020 at 3:02 PM
To: spacetelescope/jwst <jwst@noreply.github.com>
Cc: Michael Regan <mregan@stsci.edu>, Author <author@noreply.github.com>
Subject: Re: [spacetelescope/jwst] JP-1154: Flag CR neighbors #4297 (#4434)


Huh, that's not good (or normal).

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/spacetelescope/jwst/pull/4434?email_source=notifications&email_token=AF6KJLR2EZHIM7SAIAMA5B3RDWF6NA5CNFSM4KD6HEI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMJJSBY#issuecomment-588421383>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AF6KJLR2XBMS5K463PMFXJ3RDWF6NANCNFSM4KD6HEIQ>.

hbushouse commented:
On the odd chance that the failures were flukes, I've restarted a couple of the tests to see if the error repeats.

mwregan2 commented:
Howard,
  I do not see any difference between remote and local code. Is the Jenkins output saved somewhere?
Mike


From: Howard Bushouse <notifications@github.com>
Reply-To: spacetelescope/jwst <reply@reply.github.com>
Date: Wednesday, February 19, 2020 at 3:05 PM
To: spacetelescope/jwst <jwst@noreply.github.com>
Cc: Michael Regan <mregan@stsci.edu>, Author <author@noreply.github.com>
Subject: Re: [spacetelescope/jwst] JP-1154: Flag CR neighbors #4297 (#4434)


On the odd chance that the failures were flukes, I've restarted a couple of the tests to see if the error repeats.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/spacetelescope/jwst/pull/4434?email_source=notifications&email_token=AF6KJLT3XIGL6564IKO7ONLRDWGI3A5CNFSM4KD6HEI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMJJ4JA#issuecomment-588422692>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AF6KJLQJRQKEVHMXF3S6OU3RDWGI3ANCNFSM4KD6HEIQ>.

mwregan2 commented:
I knew about the "None" problem. I'm planning to change the default to be "one". It doesn't make sense to set max_cores to be "None". Meanwhile, I removed the parameters from the config file.

hbushouse commented:
I've downloaded the new code and tests and they also run OK for me, so I really can't figure out what's making them fail in Travis. They all fail at exactly the same point: right after reporting that the 5th test in `test_detect_jumps` has failed (report has "F" for that test result). Except the one test instance where we run under python 3.8.1 actually succeeded. This sounds insidious ...

jdavies-st commented:
I believe the test `test_nocr_100_groups_nframes1` is the one that is failing.  It has 100 groups and it is full frame MIRI.  That's a very large dataset in memory for a unit test, and TravisCI barfs as the memory is minimal on these VMs.  Is there a reason to have the setup be full frame when only one or a few pixels are being modified in the test?  I feel like we had this same issue when we added parallelization to this code, and most of the tests had to be paired down.

Here's the tests ordered by how long they take, which can be taken as a proxy for memory useage:

```
$ pytest jwst/jump/ --durations 10
============================================================ test session starts ============================================================
platform darwin -- Python 3.7.6, pytest-5.3.4, py-1.8.1, pluggy-0.13.1
rootdir: /Users/jdavies/dev/jwst, inifile: setup.cfg
plugins: requests-mock-1.7.0, openfiles-0.4.0, ci-watson-0.5, asdf-2.5.2a1.dev15+g90e5a4e, xdist-1.31.0, forked-1.1.3, doctestplus-0.5.0, cov-2.8.1
collected 70 items                                                                                                                          

jwst/jump/tests/test_detect_jumps.py ........s....                                                                                    [ 18%]
jwst/jump/tests/test_jump_step.py ................                                                                                    [ 41%]
jwst/jump/tests/test_twopoint_difference.py ........x..........................x.....                                                 [100%]
========================================================= slowest 10 test durations =========================================================
10.70s call     jwst/jump/tests/test_detect_jumps.py::test_nocr_100_groups_nframes1
4.26s call     jwst/jump/tests/test_detect_jumps.py::test_onecr_50_groups
1.93s call     jwst/jump/tests/test_detect_jumps.py::test_flagging_of_CRs_across_slice_boundaries
1.92s call     jwst/jump/tests/test_detect_jumps.py::test_twoints_onecr_10_groups_neighbors_flagged_multi
1.83s call     jwst/jump/tests/test_detect_jumps.py::test_twoints_onecr_each_10_groups_neighbors_flagged
1.27s call     jwst/jump/tests/test_jump_step.py::test_one_CR[None]
1.08s call     jwst/jump/tests/test_detect_jumps.py::test_crs_on_edge_with_neighbor_flagging
0.93s call     jwst/jump/tests/test_detect_jumps.py::test_onecr_10_groups_neighbors_flagged
0.89s call     jwst/jump/tests/test_detect_jumps.py::test_nocrs_noflux
0.81s call     jwst/jump/tests/test_detect_jumps.py::test_onecr_10_groups_fullarray
=========================================== 67 passed, 1 skipped, 2 xfailed, 3 warnings in 33.41s ===========================================
```

hbushouse commented:
In addition to reducing the size of the test arrays (I reduced to 100x100 and it runs a LOT faster), I'm also wondering about these lines in `setup_inputs`:
```
        times = np.array(list(range(ngroups)), dtype=np.float64) * deltatime
...
        model1 = RampModel(data=data, err=err, pixeldq=pixdq, groupdq=gdq, **times=times**)
```
There isn't an attribute called "times" in `RampModel` and it doesn't appear to be used anywhere in the tests. This should probably be removed.

hbushouse commented:
OK, the execution tests finally pass, but there are a few flake8 issues that need to be fixed (mostly unused variables). And why reduce the size of the cube for only the 1 test? Why not reduce it for all of them. There's no need to have a 1k x 1k cube when only 1-3 pixels are being used in the tests.

mwregan2 commented:
Howard,
  I thought about changing all the cubes to be smaller but some of the tests in this routine are about getting the correct result in the recombination of slices from the multiprocessing. I can’t just change the size of the cubes without working out the effect on the slice boundaries. I can do that in a future update but since the other cubes do not have 100 groups, the net savings in memory would be small.
  I saw the flake8 errors too. I can fit them now, if you want or put them in a separate pull request. Let me know which you would prefer.
Mike


From: Howard Bushouse <notifications@github.com>
Reply-To: spacetelescope/jwst <reply@reply.github.com>
Date: Thursday, February 20, 2020 at 11:03 AM
To: spacetelescope/jwst <jwst@noreply.github.com>
Cc: Michael Regan <mregan@stsci.edu>, Author <author@noreply.github.com>
Subject: Re: [spacetelescope/jwst] JP-1154: Flag CR neighbors #4297 (#4434)


OK, the execution tests finally pass, but there are a few flake8 issues that need to be fixed (mostly unused variables). And why reduce the size of the cube for only the 1 test? Why not reduce it for all of them. There's no need to have a 1k x 1k cube when only 1-3 pixels are being used in the tests.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/spacetelescope/jwst/pull/4434?email_source=notifications&email_token=AF6KJLSSUMMLZPHMLJMLJUTRD2SUFA5CNFSM4KD6HEI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMO4W2A#issuecomment-589155176>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AF6KJLSBUUXPXC24NPI65UTRD2SUFANCNFSM4KD6HEIQ>.

hbushouse commented:
I'd like to see the flake8 issues fixed before we merge this. Otherwise every build is going to generate those errors.

jdavies-st commented:
Multiprocessing should be breaking them up into chunks, and as long as the cubes are larger than the chunks, it should be fine, right?  The cubes being too large was an issue with the original PR implementing multiprocessing last summer.  And we fixed it then by having many of the fixtures produce smaller cubes.  It's not clear to me why the need to be larger now for all the new tests.

It would be good to split-out tests that are testing multiprocessing and need a full frame from those that are not testing multiprocessing and do not.  That latter can run *much* faster on a 10x10 or 50x50 subarray (or be able to run at all, especially if `ngroups` is large).  There can be separate fixtures for each of these.

mwregan2 commented:
The CR neighbor flagging changes mean that pixels on the first or last row in a slice now can cause a pixel on the previous or next slice to be flagged. So, the code now keeps track of any jumps on the first or last row of a slice and in the recombination process adds the flags to the adjacent slices. To do test this II need to know how many rows are in each slice which depends on the number of cores and the number of rows. So, I would have to make changes to the tests. I can make the cube smaller but  those tests are only 10 group integrations. So, the savings in memory would be around 40 MB or so. Changing the 100 group integrations to be 100x100 was a savings of 400 MB.
  If you want, I can and will make the change. It is not just a simple change in the size though. I have to go back and work what the slice row sizes should be. Any new calculation means there’s a chance I’ll do it wrong. I just feel that the time required, right now, is not worth the gain. But it is your call, If you want me to do it before the merge I will.
Mike


From: James Davies <notifications@github.com>
Reply-To: spacetelescope/jwst <reply@reply.github.com>
Date: Thursday, February 20, 2020 at 11:27 AM
To: spacetelescope/jwst <jwst@noreply.github.com>
Cc: Michael Regan <mregan@stsci.edu>, Author <author@noreply.github.com>
Subject: Re: [spacetelescope/jwst] JP-1154: Flag CR neighbors #4297 (#4434)


Multiprocessing should be breaking them up into chunks, and as long as the cubes are larger than the chunks, it should be fine, right? The cubes being too large was an issue with the original PR implementing multiprocessing last summer. And we fixed it then by having many of the fixtures produce smaller cubes. It's not clear to me why the need to be larger now for all the new tests.

It would be good to split-out tests that are testing multiprocessing and need a full frame from those that are not testing multiprocessing and do not. That latter can run much faster on a 10x10 or 50x50 subarray (or be able to run at all, especially if ngroups is large). There can be separate fixtures for each of these.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/spacetelescope/jwst/pull/4434?email_source=notifications&email_token=AF6KJLW43XH5Z3TWRTH4KGDRD2VNZA5CNFSM4KD6HEI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMO7TQQ#issuecomment-589167042>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AF6KJLWS4LZNQFN47WI6DBLRD2VNZANCNFSM4KD6HEIQ>.

jdavies-st commented:
I think it would probably be better to have the default size of the input arrays be a small subarray, as most tests only need that.  If a test needs to know the size for dealing with edge issues, the test itself should then specify `nrows` and ncols` when calling `setup_inputs()`, and then also use those `nrows` and `ncols` in the test itself to populated CRs and do the asserts.  This would be easier to maintain for the future and clearer in the test what is happening.

And for multiprocessing tests that need a full array, or an array of some size or larger, again that test can specify.  Have an assert in the test that makes sure at least two slices are being generated regardless of the input array size to the test.  Then you can be certain it is testing what you think it is testing.  Better not to have logic in a test, i.e. `if numslices > 1:`, because that means that test is not doing anything on a system where there are not at least 2 cores available for whatever reason.  Logic in tests is not good.  Better to have lots of asserts so you know exactly what it is doing and can catch bad behavior when it is not doing what you expect.  Better to have `assert numslices > 1` in the test itself.

mwregan2 commented:
I agree that the default size should be a smaller array. It also takes work to make the change. Not much work,  but non-zero. The gain is not clear to me. Effectively we'd only be testing subarrays. It should not be a problem but there exist bugs that would only be found with full size arrays. They would be pretty weird but are possible. 
Wouldn't having an assert of numslices > 1  cause a test failure when nothing is wrong in the code and we are just in an unexpected environment? I don't think it is incorrect to run the pipeline on a single core single thread machine. While I agree, in principle, with no logic in tests, it seems like skipping a test when the environment does not allow it to run  correctly makes sense.

hbushouse commented:
I'm OK with merging this as-is, now that it at least runs and the longest-running test has been sped up. We can always revisit the optimization of the tests later.

jdavies-st commented:
Sounds good Howard.  We can streamline the tests in a future PR.

For the tests that can't run due to a single core available, it might make sense to `@pytest.mark.skipif()` them for that reason, then it shows up in the output that they didn't run.  Currently the test would pass silently because no asserts are even happening.
